compiler/z/codegen/OMRMachine.cpp

```diff
@@ -5327,40 +5327,70 @@ OMR::Z::Machine::coerceRegisterAssignment(TR::Instruction
          }
       else
          {
-         self()->cg()->traceRegAssigned(currentTargetVirtual, spareReg);
-
-         // virtual register is not assigned yet, copy register
-         cursor = self()->registerCopy(currentInstruction, currentTargetVirtualRK, targetRegister, spareReg, self()->cg(), instFlags);
-
-         spareReg->setState(TR::RealRegister::Assigned);
-         spareReg->setAssignedRegister(currentTargetVirtual);
-         currentTargetVirtual->setAssignedRegister(spareReg);
+         // if spareReg is still NULL, then it means we did not find any free Reg, and no
+         // other register is available to spill. So we must spill the target register.
+
+         // In general if a register is blocked, the virtual register
+         // associated with a target register would have been assigned as part of that instruction.
+         // So we would not do this (i.e. spill the blocked register). However, in this case all
+         // pre-assigned registers in the dependency are blocked before individiual dependencies are coerced
+         // into their respective target registers (inside TR_S390RegisterDependencyGroup::assignRegisters((..)).
+         // In this situation it's possible that the target register of a register dependency is currently
+         // blocked (as in here) and occupied by another virtual register and that there is no other available
+         // register to free.
+         // Hence, we must spill the blocked target reg as it is the only available candidate.
+         if (spareReg == NULL)
+            {
+            self()->cg()->setRegisterAssignmentFlag(TR_RegisterSpilled);
+            virtualRegister->block();
```

Comment:
`currentAssignedRegister` is `NULL` at this point so why do we need the block `virtualRegister` here? It should not be associated with any real register so blocking doesn't do anything.